linkerd/service-profiles/src/http/service.rs

```diff
@@ -0,0 +1,379 @@
+            }),
+        }
+    }
+}
+
+impl<T, G, R, C> tower::Service<T> for MakeSvc<G, R, C, SmallRng>
+where
+    T: WithRoute + Clone,
+    G: GetRoutes<T>,
+    R: NewService<T::Route> + Clone,
+    C: Clone,
+{
+    type Response = Service<T, R, Override<C>>;
+    type Error = Error;
+    type Future = MakeFuture<T, G::Future, R, C, SmallRng>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.get_routes.poll_ready().map_err(Into::into)
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let future = self.get_routes.get_routes(target.clone());
+
+        MakeFuture {
+            future,
+            inner: Some(Inner {
+                target,
+                make_route: self.make_route.clone(),
+                default_route: self.default_route.clone(),
+                make_concrete: self.make_concrete.clone(),
+                dst_override: self.dst_override.clone(),
+            }),
+        }
+    }
+}
+
+impl<T, F, R, C> Future for MakeFuture<T, F, R, C, ()>
+where
+    T: WithRoute + Clone,
+    F: Future<Item = watch::Receiver<Routes>>,
+    F::Error: Into<Error>,
+    R: NewService<T::Route>,
+{
+    type Item = Service<T, R, Forward<C>>;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        trace!("poll");
+        let profiles = try_ready!(self.future.poll().map_err(Into::into));
+
+        let Inner {
+            target,
+            make_route,
+            default_route,
+            make_concrete,
+            dst_override: (),
+        } = self.inner.take().unwrap();
+
+        let requests = Requests::new(target.clone(), make_route, default_route);
+        let svc = Service {
+            profiles,
+            requests,
+            concrete: Forward(make_concrete),
+        };
+
+        trace!("forwarding profile service ready");
+        Ok(svc.into())
+    }
+}
+
+impl<T, F, R, C> Future for MakeFuture<T, F, R, C, SmallRng>
+where
+    T: WithRoute + Clone,
+    F: Future<Item = watch::Receiver<Routes>>,
+    F::Error: Into<Error>,
+    R: NewService<T::Route> + Clone,
+{
+    type Item = Service<T, R, Override<C>>;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        trace!("poll");
+        let profiles = try_ready!(self.future.poll().map_err(Into::into));
+
+        let Inner {
+            target,
+            make_route,
+            default_route,
+            make_concrete,
+            dst_override: rng,
+        } = self.inner.take().unwrap();
+
+        let requests = Requests::new(target.clone(), make_route, default_route);
+        let (service, update) = concrete::default(make_concrete, rng);
+        let svc = Service {
+            profiles,
+            requests,
+            concrete: Override { service, update },
+        };
+
+        trace!("overriding profile service ready");
+        Ok(svc.into())
+    }
+}
+
+mod sealed {
+    pub trait PollUpdate {
```

Comment:
it would be nice if there was a comment here explaining what this is for...its purpose is not very obvious.